### PR TITLE
Implement `cuda::mr::managed_memory_resource`

### PR DESCRIPTION
**Pull request by miscco**
Fixes Implement a memory_resource using `cudaMallocManaged` and `cudaFree` #1515

Respin of #1556 due to stacked branch issues

### CODE REVIEW

File: libcudacxx/include/cuda/__memory_resource/cuda_managed_memory_resource.h [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+public:
+  constexpr cuda_managed_memory_resource(const unsigned int __flags = cudaMemAttachGlobal) noexcept
+      : __flags_(__flags & __available_flags)
+  {
+    _LIBCUDACXX_ASSERT(__flags_ == __flags, "Unexpected flags passed to cuda_managed_memory_resource");
+  }
+
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_managed_memory_resource.h:59]: Suggestion: It's only device memory the first time it is accessed on the device (or based on cudaMemAdvise hints applied to a range of it). 

```suggestion
   * @brief Allocate CUDA unified memory of size at least \p __bytes.
```

File: libcudacxx/include/cuda/__memory_resource/cuda_managed_memory_resource.h [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation.
+   * @throw cuda::cuda_error of the returned error code
+   * @return void* Pointer to the newly allocated memory
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_managed_memory_resource.h:63]: Suggestion: I think doxygen automates documenting the return type, so putting it in the docstring is redundant.

```suggestion
   * @return Pointer to the newly allocated memory
```

File: libcudacxx/include/cuda/__memory_resource/properties.h [hunk 3]
```diff
@@ -42,6 +42,11 @@ struct device_accessible
 struct host_accessible
 {};
 
+/// \struct managed_memory
+/// \brief The \c managed_memory property signals that the allocated memory is managed
+struct managed_memory
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/properties.h:47]: I think this change requires changes to the documentation. 
https://nvidia.github.io/cccl/libcudacxx/extended_api/memory_resource/properties.html

Specifically: 

> For now, libcu++ provides two commonly used properties: `cuda::mr::device_accessible` and `cuda::mr::host_accessible`.

File: libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_managed_memory_resource/equality.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                                   cuda::mr::managed_memory>,
+    "");
+static_assert(
+    cuda::mr::async_resource<async_resource<AccessibilityType::Device> >, "");
+static_assert(
+    cuda::mr::async_resource_with<async_resource<AccessibilityType::Device>,
+                                  cuda::mr::managed_memory>,
+    "");
+
+// Ensure that we can only
```
Comment (harrism) [inline libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_managed_memory_resource/equality.pass.cpp:66]: Incomplete comment (I saw this in another PR as well -- one that got closed?).

**Review by harrism — CHANGES_REQUESTED** [review]
Looks mostly good to go. A few suggestions.

File: libcudacxx/docs/extended_api/memory_resource/properties.md [hunk 5]
```diff
@@ -31,9 +31,14 @@ template<class MemoryResource, class = cuda::std::enable_if_t<cuda::has_property
...
+For now, libcu++ provides various commonly used properties:
 
-Note that the libcu++ provided properties are stateless. However, properties can also provide stateful information that is retrieved via the `get_property` free function. In order to communicate the desired type of the carried state, a stateful property must define the `value_type` alias. A library can constrain interfaces that require a stateful property with `cuda::has_property_with` as shown in the example below (See [here](https://godbolt.org/z/11sGbr333) for a minimal Compiler Explorer example).
+* `cuda::mr::device_accessible` and `cuda::mr::host_accessible` indicate whether memory allocated using a memory resource is accessible from host or device respectively.
+* `cuda::mr::managed_memory` indicates that the memory resource allocates CUDA unified memory which is both host and device accessible
+
+More properties may be added as the library and the hardware capabilities evolve. However, a user library is free to define as many properties as needed to fully cover its API surface.
```
Comment (harrism) [inline libcudacxx/docs/extended_api/memory_resource/properties.md:39]: Suggestion: simplify, so you don't imply libraries typically need a plethora of properties.

```suggestion
More properties may be added as the library and the hardware capabilities evolve. However, a user library is free to define custom properties.
```

File: libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_managed_memory_resource/allocate.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+// UNSUPPORTED: nvrtc
+
+#include <cuda/memory_resource>
+#include <cuda/std/cassert>
+#include <cuda/std/cstdint>
+#include <cuda/stream_ref>
+
+#include "test_macros.h"
+
+void ensure_device_ptr(void* ptr) {
```
[the remark this answers is not in the document]
Comment (miscco) [reply libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_managed_memory_resource/allocate.pass.cpp:22]: I am not sure, in the end we test different functionality and it makes it easier to see what is tested if the function is local

**Review by harrism — APPROVED** [review]

**Review by griwes — APPROVED** [review]

**Pull request merged**